// firadio/index.js
/*
 * firhttp开源项目官方网址http://node.firadio.com
 * 特点：简单、快速、无压缩、无缓存、只处理动态页面、不放图片/JS/CSS
 * 如果修改代码请保留注释，并通过SVN提交上来.
 * SVN地址https://firadionode.googlecode.com/svn
 * SVN帐号密码获取请联系作者QQ:550220138获取
 * 如果您对firhttp的研究感兴趣，可以加入我们的QQ群26499890
*/
module.exports.date=require('./date').date;
module.exports.gbk_to_utf8=require('./gbk_to_utf8').gbk_to_utf8;
module.exports.log=(function(str){
	console.log(module.exports.date('H:i:s.K')+' '+str);
});
module.exports.log1=(function(str){//开始
	process.stdout.write(module.exports.date('H:i:s.K')+' '+str);
});
module.exports.log2=(function(str){//中间
	process.stdout.write(str);
});
module.exports.log3=(function(str){//结束
	console.log(str);
});
module.exports.log4=(function(str){//执行过程需要报错
	console.log("\r\n"+str);
});
module.exports.http=require('./http');
module.exports.dns=require('./dns');
module.exports.base64=require('./base64');
module.exports.intval=(function(mixed_var, base){
    var tmp;
    var type = typeof(mixed_var);
     if (type === 'boolean') {
        return +mixed_var;
    } else if (type === 'string') {
        tmp = parseInt(mixed_var, base || 10);
        return (isNaN(tmp) || !isFinite(tmp)) ? 0 : tmp;    } else if (type === 'number' && isFinite(mixed_var)) {
        return mixed_var | 0;
    } else {
        return 0;
    }
});
module.exports.md5=require('./md5').md5;
